qubes-gui-client/src/buffer.rs

```diff
@@ -0,0 +1,389 @@
+}
+
+impl VchanMock for vchan::Vchan {
+    fn buffer_space(&self) -> usize {
+        vchan::Vchan::buffer_space(self)
+    }
+    fn recv(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        vchan::Vchan::recv(self, buf)
+    }
+    fn send(&mut self, buf: &[u8]) -> io::Result<usize> {
+        vchan::Vchan::send(self, buf)
+    }
+    fn wait(&self) {
+        vchan::Vchan::wait(self)
+    }
+    fn data_ready(&self) -> usize {
+        vchan::Vchan::data_ready(self)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Vchan<T: VchanMock> {
+    vchan: T,
+    queue: VecDeque<Vec<u8>>,
+    offset: usize,
+    state: ReadState,
+    header: Header,
+    buffer: Vec<u8>,
+}
+
+#[inline(always)]
+fn u32_to_usize(i: u32) -> usize {
+    // If u32 doesn’t actually fit in a usize, fail the build
+    let [] = [0; if u32::MAX as usize as u32 == u32::MAX {
+        0
+    } else {
+        1
+    }];
+    i as usize
+}
+
+impl<T: VchanMock> Vchan<T> {
+    fn write_slice(vchan: &mut T, slice: &[u8]) -> io::Result<usize> {
+        let space = vchan.buffer_space();
+        if space == 0 {
+            Ok(0)
+        } else {
+            let to_write = space.min(slice.len());
+            vchan.send(&slice[..to_write])
+        }
+    }
+
+    fn drain(&mut self) -> io::Result<usize> {
+        let mut written = 0;
+        loop {
+            let front: &mut _ = match self.queue.front_mut() {
+                None => break Ok(written),
+                Some(e) => e,
+            };
+            let to_write = &front[self.offset..];
+            if to_write.is_empty() {
+                self.queue.pop_front();
+                self.offset = 0;
+                continue;
+            }
+            let written_this_time = Self::write_slice(&mut self.vchan, to_write)?;
+            written += written_this_time;
+            self.offset += written_this_time;
+            if written_this_time < to_write.len() {
+                break Ok(written);
+            }
+        }
+    }
+
+    pub fn write(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.drain()?;
+        if !self.queue.is_empty() {
+            self.queue.push_back(buf.to_owned());
+            return Ok(());
+        }
+        assert_eq!(self.offset, 0);
+        let written = Self::write_slice(&mut self.vchan, buf)?;
+        if written != buf.len() {
+            assert!(written < buf.len());
+            self.queue.push_back(buf[written..].to_owned());
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn recv(&mut self, s: Range<usize>) -> io::Result<usize> {
+        self.vchan.recv(&mut self.buffer[s]).map_err(|e| {
+            self.state = ReadState::Error;
+            e
+        })
+    }
+
+    pub fn wait(&mut self) {
+        self.vchan.wait()
+    }
+
+    /// If there is nothing to read, return `Ok(None)` immediately; otherwise,
+    /// returns `Ok(Some(msg))` if a complete message has been read, or `Err`
+    /// if something went wrong.
```

Comment:
This doesn't actually specify what happens if there is a partial message.

```suggestion
    /// If a complete message is available in the buffer, return `Ok(Some(msg))`; otherwise more data
    /// still needs to arrive, which is signaled by returning `Ok(None)`. If an error occurs, `Err` is returned.
```

What should the user do if the error occurs? What does it say about the state of our buffers, and the state of the protocol?

Reply:
I’m not sure how an error should be handled tbh.  Should it cause disconnection?

Reply:
Error while reading from vchan? Yes, I think disconnect (and listen for reconnect) is appropriate reaction.

Reply:
Malformed message from vchan.